daemon/list.go

```diff
@@ -133,7 +133,18 @@ func (daemon *Daemon) Containers(config *ContainersConfig) ([]*types.Container,
 			ID:    container.ID,
 			Names: names[container.ID],
 		}
-		newC.Image = container.Config.Image
+
+		img, err := daemon.Repositories().LookupImage(container.Config.Image)
+		if err != nil {
+			// If the image can no longer be found by its original reference,
+			// it makes sense to show the ID instead of a stale reference.
+			newC.Image = container.ImageID
```

Comment:
Just noticed that I missed a question in your response @thieman. The second example you ask about is what is stored internally. The problem is that the code was still applying ":latest" because the image was originally started as "nginx:latest" on the command line.  Here's the actual Docker 1.6.0 output from that same container:

```
d419fa7ad1a7        2485b0f8995167b796a7ab4696e842c08bd27d58febb3ba0e935a4e2c1b2807d:latest   "nginx -g 'daemon of   9 weeks ago         Up 8 seconds        443/tcp, 0.0.0.0:10234->80/tcp   dockercon-6c01869525db08
```